src/compiler/checker.ts

```diff
@@ -44504,9 +44504,35 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return PredicateSemantics.Sometimes;
                 }
                 return PredicateSemantics.Always;
+            case SyntaxKind.BigIntLiteral:
+                if ((node as BigIntLiteral).text === "0n") {
+                    // unlike `while(0)`, `while (0n)` is not idiomatic.
+                    return PredicateSemantics.Never;
+                }
+                return PredicateSemantics.Always;
+            // handle +123, -123, -123n, etc.
+            case SyntaxKind.PrefixUnaryExpression:
+                const prefixUnaryExpression = node as PrefixUnaryExpression;

```

Comment:
Seems like this could just be something like
```ts
if (operand.kind === SyntaxKind.NumericLiteral && (operator === SyntaxKind.MinusToken || operator === SyntaxKind.PlusToken)) {
  return getTruthySemantics(operand);
}
```
? We shouldn't be duplicating all the logic of the numeric case.

Reply:
Fair question. I wrote it this way to avoid a few bugs....
```ts
if (1) {} // this is intentionally exempted from the check, BUT
if (-1) {} // should report. Would NOT with the recursion.
if (+1) {} // IMO should also report, since this is clearly intended as a number, not as a shorthand for `if (true)`.

// (ditto for the following)
if (0) {} // intentionally exempted from reporting
if (+0) {} // should report IMO. This is clearly intended as a number, and not as a shorthand for `if (false)`.
if (-0) {} // see previous.
```

Note that the branches are therefore different:
```ts
            case SyntaxKind.NumericLiteral:
                // Allow `while(0)` or `while(1)`
                if ((node as NumericLiteral).text === "0" || (node as NumericLiteral).text === "1") {
                    return PredicateSemantics.Sometimes; // <-- special case; while(0) or while(1).
                }
                return PredicateSemantics.Always;
```
vs
```ts
                if (operand.kind === SyntaxKind.NumericLiteral && (operator === SyntaxKind.MinusToken || operator === SyntaxKind.PlusToken)) {
                    if ((operand as NumericLiteral).text === "0") {
                        return PredicateSemantics.Never; // <-- just a numeric zero
                    }
                    else {
                        return PredicateSemantics.Always;
                    }
                }
```

I _think_ the bigint case is ok to recurse on the unary `-` operator. However, we just need to be sure it doesn't get hit by recursion from the unary `+` operator, since `if (+0n)` isn't falsy. It's an exception!

FWIW, note that [ESLint does recurse in the bigint unary `+` case](https://github.com/eslint/eslint/blob/3a4eaf921543b1cd5d1df4ea9dec02fab396af2a/lib/rules/utils/ast-utils.js#L949), and therefore [does incur this bug](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tY29uc3RhbnQtYmluYXJ5LWV4cHJlc3Npb246IFwiZXJyb3JcIiovXG4vKmVzbGludCBuby1jb25zdGFudC1jb25kaXRpb246IFwiZXJyb3JcIiovXG5cbmlmICgrM24pIHt9Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJzb3VyY2VUeXBlIjoibW9kdWxlIiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QifX19).

So between 

a) recurse in these cases, and add context to the recursion to be able to handle the edge cases.
b) recurse in this case, and accept some IMO faulty edge cases
c) Not recurse at all

I chose c) for simplicity, and added test cases around the `while (+0)`, etc. cases.

But LMK if you'd like a different approach! 

Note that within c) there's one opportunity that I see for deduping, which is to combine the bigint checks.
```ts
        function getBigIntLiteralTruthySemantics (node: BigIntLiteral): PredicateSemantics {
            return node.text === "0n" ? PredicateSemantics.Never : PredicateSemantics.Always;

        }

        switch (node.kind) {
            case SyntaxKind.BigIntLiteral:
                    // unlike `while(0)`, `while (0n)` is not idiomatic.
                return getBigIntLiteralTruthySemantics(node as BigIntLiteral);
            // handle +123, -123, -123n, etc.
            case SyntaxKind.PrefixUnaryExpression:
                const prefixUnaryExpression = node as PrefixUnaryExpression;
                const { operator, operand } = prefixUnaryExpression;
                if (operand.kind === SyntaxKind.BigIntLiteral && operator === SyntaxKind.MinusToken) {
                    return getBigIntLiteralTruthySemantics(node as BigIntLiteral);
                } else 
                // etc
```